### ISSUE TITLE
Value Checker cannot find class for @StaticallyExecutable on static methods

### ISSUE THREAD

**Issue by ThiagoTGM** [typetools/checker-framework#3630]
Trying to use the `@StaticallyExecutable` on a static method doesn't seem to work, it always fails to find the class.
The class I'm trying to compile is:

```java
package com.github.thiagotgm.forester;

import org.checkerframework.common.value.qual.IntVal;
import org.checkerframework.common.value.qual.StaticallyExecutable;
import org.checkerframework.dataflow.qual.Pure;

public class App {
    public static void main( String[] args ) {
        System.out.println( "Hello World!" );
        new App().bar();
    }
    
    @StaticallyExecutable @Pure
    public static int myAdd( int a, int b ) {
        return a + b;
    }
    
    public void bar() {
        @IntVal({ 1 }) int a = 1;
        @IntVal({ 2 }) int b = 2;
        @IntVal({ 3 }) int v = myAdd( a, b );
    }
}
```

Before compiling with Checker, I used Maven to generate a jar of the project (`mvn package`), copied the jar to the project basedir (`/mnt/c/Users/thiag/GitRepositories/Forester`), then added that jar to the classpath. The compiler command I used is:

```bash
javac -d /mnt/c/Users/thiag/GitRepositories/Forester/target/classes -classpath /mnt/c/Users/thiag/GitRepositories/Forester/target/classes:/home/thiagotgm/.m2/repository/org/checkerframework/checker-qual/3.6.1/checker-qual-3.6.1.jar:/mnt/c/Users/thiag/GitRepositories/Forester/forester-0.0.1-SNAPSHOT.jar: -sourcepath /mnt/c/Users/thiag/GitRepositories/Forester/src/main/java:/mnt/c/Users/thiag/GitRepositories/Forester/target/generated-sources/annotations: -s /mnt/c/Users/thiag/GitRepositories/Forester/target/generated-sources/annotations -processor org.checkerframework.common.value.ValueChecker -processorpath /home/thiagotgm/.m2/repository/org/checkerframework/checker/3.6.1/checker-3.6.1.jar: -g --release 11 -encoding UTF-8 -Xmaxerrs 10000 -Xmaxwarns 10000 -AreportEvalWarns -AprintAllQualifiers -version -verbose src/main/java/com/github/thiagotgm/forester/App.java
```

(Note: Also tried to set the .class output itself by copying `./target/classes` to basedir then replacing the jar with that directory in the classpath, still didn't work)
(Note 2: The manual said to also include a `-AprintGitProperties` option, but that causes it to [throw an exception](https://pastebin.com/CsyGRfZ6))

I expected that the checker would load the class from the classpath, run it, and compile successfully, as this is basically the example given in [section 21.2.2](https://checkerframework.org/manual/#other-constant-value-annotations) of the manual, but with the annotated method being `static` instead of an instance method (which isn't disallowed in the manual or the javadocs).

Instead, the checker gives a `[class.find.failed] Failed to find class named com.github.thiagotgm.forester.App` warning, before failing the compile due to an incompatible type (`@UnknownVal` to `@IntVal`). See the full log [here](https://pastebin.com/PGXDAnai).

Changing the code to be effectively the same as the manual example (i.e. removing the `static` from `myAdd`) makes the `class.find.failed` warning go away, but it still fails compilation with an incompatible type error (see [this log](https://pastebin.com/5iYFZBt8)).

OS: Ubuntu 20.04 (WSL2)
Compiler Version: 11.0.8 (openjdk-11)
Target Version: 11
Checker Version: 3.6.1

**Comment by mernst** [comment]
I'm sorry you are having trouble.  Thanks for reporting the problem.

Unfortunately, I cannot reproduce your issue.

Here are the commands that I ran:

```
# Must use JDK 11!
mkdir ~/tmp
cd ~/tmp
rm -rf Forester
git clone https://github.com/ThiagoTGM/Forester
cd Forester
Forester="$(pwd -P)"
mvn package
```

At this point there is no `${Forester}/forester-0.0.1-SNAPSHOT.jar` file, which is used in the command you gave.  So, there must be some difference between the commands that you and I ran.

Could you please give commands that can be cut-and-pasted in order to reproduce your problem?  (You could start with the commands I gave above, if that is helpful.)  Then we can investigate and fix the problem.  Thanks!

**Comment by ThiagoTGM** [comment]
I copied `forester-0.0.1-SNAPSHOT.jar` manually from `target`, sorry if that wasn't clear. I was doing that due to a separate issue with Maven and I forgot to change it here, sorry. Here's the fixed command:

```bash
javac -d /mnt/c/Users/thiag/GitRepositories/Forester/target/classes -classpath /mnt/c/Users/thiag/GitRepositories/Forester/target/classes:/home/thiagotgm/.m2/repository/org/checkerframework/checker-qual/3.6.1/checker-qual-3.6.1.jar:/mnt/c/Users/thiag/GitRepositories/Forester/target/forester-0.0.1-SNAPSHOT.jar: -sourcepath /mnt/c/Users/thiag/GitRepositories/Forester/src/main/java:/mnt/c/Users/thiag/GitRepositories/Forester/target/generated-sources/annotations: -s /mnt/c/Users/thiag/GitRepositories/Forester/target/generated-sources/annotations -processor org.checkerframework.common.value.ValueChecker -processorpath /home/thiagotgm/.m2/repository/org/checkerframework/checker/3.6.1/checker-3.6.1.jar: -g --release 11 -encoding UTF-8 -Xmaxerrs 10000 -Xmaxwarns 10000 -AreportEvalWarns -AprintAllQualifiers -version -verbose src/main/java/com/github/thiagotgm/forester/App.java
```

**Comment by mernst** [comment]
Running these commands:

```
# Must use JDK 11!
mkdir ~/tmp
cd ~/tmp
rm -rf Forester
git clone https://github.com/ThiagoTGM/Forester
cd Forester
Forester="$(pwd -P)"
mvn package
javac -d ${Forester}/target/classes -classpath ${Forester}/target/classes:${HOME}/.m2/repository/org/checkerframework/checker-qual/3.6.1/checker-qual-3.6.1.jar:${Forester}/target/forester-0.0.1-SNAPSHOT.jar: -sourcepath ${Forester}/src/main/java:${Forester}/target/generated-sources/annotations: -s ${Forester}/target/generated-sources/annotations -processor org.checkerframework.common.value.ValueChecker -processorpath ${HOME}/.m2/repository/org/checkerframework/checker/3.6.1/checker-3.6.1.jar: -g --release 11 -encoding UTF-8 -Xmaxerrs 10000 -Xmaxwarns 10000 -AreportEvalWarns -AprintAllQualifiers -version -verbose src/main/java/com/github/thiagotgm/forester/App.java
```

gives me

```
javac 11.0.2
error: Annotation processor 'org.checkerframework.common.value.ValueChecker' not found
[loading /home/mernst/tmp/Forester/target/classes/module-info.class]
error: module not found: org.checkerframework.checker.qual
[loading modules/java.base/module-info.class]
[total 37ms]
2 errors
```

The "error: Annotation processor ... not found" error is because you have misconfigured the processorpath.  You set it to 

```
-processorpath ${HOME}/.m2/repository/org/checkerframework/checker/3.6.1/checker-3.6.1.jar:
```

but that path does not exist, after following the instructions above.  So you must have done something else to create that file.

Could you please give complete instructions that can be cut-and-pasted into a shell, and that work for an arbitrary user?  They should never mention hard-coded paths such as your own home directory.
You could try creating a new user account and running them from that user account, then paste the full set of commands into this issue.  Thanks!

**Comment by ThiagoTGM** [comment]
Ah, right. It's because the dependency with the checkers doesn't get resolved without using a profile. Oops. Updated that in the pom in a test branch to get that fixed.

Anyway, these commands now work on a brand new Ubuntu VM (after setting up JDK11 and Maven):
```bash
mkdir ~/tmp
cd ~/tmp
rm -rf Forester
git clone https://github.com/ThiagoTGM/Forester
cd Forester
git checkout checker-test
Forester="$(pwd -P)"
mvn package
javac -d ${Forester}/target/classes -classpath {Forester}/target/classes:${HOME}/.m2/repository/org/checkerframework/checker-qual/3.6.1/checker-qual-3.6.1.jar:${Forester}/target/forester-0.0.1-SNAPSHOT.jar: -sourcepath ${Forester}/src/main/java:${Forester}/target/generated-sources/annotations: -s ${Forester}/target/generated-sources/annotations -processor org.checkerframework.common.value.ValueChecker -processorpath ${HOME}/.m2/repository/org/checkerframework/checker/3.6.1/checker-3.6.1.jar: -g --release 11 -encoding UTF-8 -Xmaxerrs 10000 -Xmaxwarns 10000 -AreportEvalWarns -AprintAllQualifiers -version -verbose ./src/main/java/com/github/thiagotgm/forester/App.java
```

**Comment by mernst** [comment]
Thank you.  Using the instructions you gave, I can reproduce the `[class.find.failed] Failed to find class named com.github.thiagotgm.forester.App.` warning.

**Comment by mernst** [comment]
The problem is that you used `-processorpath` but but the classes on the classpath.  If you use the processorpath, then the Checker Framework only has access to the processor path.  If you don't use the processorpath, then the Checker Framework has access to the classpath.  This Java detail is a bit obscure, and was not clearly explained in the manual.

Here are commands that work for me.  They are the same as yours except adjusting the classpath and processorpath.

```
mkdir ~/tmp
cd ~/tmp
rm -rf Forester
git clone https://github.com/ThiagoTGM/Forester
cd Forester
git checkout checker-test
Forester="$(pwd -P)"
mvn package
javac -d ${Forester}/target/classes -classpath ${HOME}/.m2/repository/org/checkerframework/checker-qual/3.6.1/checker-qual-3.6.1.jar -sourcepath ${Forester}/src/main/java:${Forester}/target/generated-sources/annotations: -s ${Forester}/target/generated-sources/annotations -processor org.checkerframework.common.value.ValueChecker -processorpath ${HOME}/.m2/repository/org/checkerframework/checker/3.6.1/checker-3.6.1.jar:${Forester}/target/classes:${Forester}/target/forester-0.0.1-SNAPSHOT.jar -g --release 11 -encoding UTF-8 -Xmaxerrs 10000 -Xmaxwarns 10000 -AreportEvalWarns -AprintAllQualifiers -version  ./src/main/java/com/github/thiagotgm/forester/App.java
```

**Comment by ThiagoTGM** [comment]
I see. Moving the self-dependency to the processorPath in the Maven config seems to have worked so you're right (with the added bonus that apparently maven apparently allows actual self-dependencies in the processor path rather than needing workarounds with `system` dependencies).

That said, it seems that removing the `static` modifier from `myAdd` causes it to become a `assignment.type.incompatible` error, despite being exactly the same as the example in the manual (it doesn't seem to be executing the method as it says `myAdd( a, b )` is `@UnknownVal`). No warnings about not finding the class or method are given. Should I open a separate issue for this?

**Comment by mernst** [comment]
`@StaticallyExecutable` is not generally applicable to non-static methods.
When `myAdd` is an instance method, it takes three arguments, whose types are `App`, `int`, and `int`.  The source code `myAdd(a, b)` is actually shorthand for `this.myAdd(a, b)`.  In that invocation, the first argument `this` is not a compile-time constant, so the invocation cannot be statically executed.

The manual is incorrect in that the example lacks `static`.  I apologize for that!  I have fixed it.

There are some bugs here that deserve issues (in addition to the documentation problems that are fixed in pull request #3647 -- thank you for prompting those fixes!).

 * The Constant Value Checker should warn if a @StaticallyExecutable method is not `@Pure`.
 * The Constant Value Checker should warn if a @StaticallyExecutable method has a formal parameter (including the receiver) that cannot be a compile-time constant.

I opened #3648 to track them.
